File: flutter_app/lib/entities/game_entities.dart
import '../core/palette.dart';

class HitboxRect {
  final double x;
  final double y;
  final double w;
  final double h;

  HitboxRect(this.x, this.y, this.w, this.h);

  bool intersects(HitboxRect other) {
    return x < other.x + other.w &&
        x + w > other.x &&
        y < other.y + other.h &&
        y + h > other.y;
  }
}

class RickshawPlayer {
  int lane = 1;
  late double x;
  late double y;
  late double startX;
  late double targetX;
  bool isChangingLane = false;
  double laneChangeProgress = 1;

  int lives = 3;
  int maxLives = 3;
  double invincibleTimerSec = 0;

  bool isTurbo = false;
  double turboTimeSec = 0;

  bool isShielded = false;

  bool isMagnetActive = false;
  double magnetTimeSec = 0;

  int nearMissCount = 0;

  RickshawPlayer() {
    _initPosition();
  }

  void _initPosition() {
    final laneCenters = [50.0, 90.0, 130.0];
    x = laneCenters[1] - 16;
    y = 232;
    startX = x;
    targetX = x;
  }

  void startLaneChange(int newLane) {
    final laneCenters = [50.0, 90.0, 130.0];
    if (newLane >= 0 && newLane < 3 && newLane != lane) {
      lane = newLane;
      startX = x;
      targetX = laneCenters[lane] - 16;
      isChangingLane = true;
      laneChangeProgress = 0;
    }
  }

  void update(double dt) {
    if (isChangingLane) {
      laneChangeProgress += dt / 0.14;
      if (laneChangeProgress >= 1) {
        laneChangeProgress = 1;
        isChangingLane = false;
        x = targetX;
      } else {
        x = startX + (targetX - startX) * laneChangeProgress;
      }
    }

    if (isTurbo) {
      turboTimeSec -= dt;
      if (turboTimeSec <= 0) {
        isTurbo = false;
        turboTimeSec = 0;
      }
    }

    if (isMagnetActive) {
      magnetTimeSec -= dt;
      if (magnetTimeSec <= 0) {
        isMagnetActive = false;
        magnetTimeSec = 0;
      }
    }

    if (invincibleTimerSec > 0) {
      invincibleTimerSec -= dt;
      if (invincibleTimerSec < 0) invincibleTimerSec = 0;
    }
  }

  HitboxRect getHitbox() {
    return HitboxRect(x + 6, y + 16, 20, 24);
  }
}

class ObstacleEntity {
  double x = 0;
  double y = 0;
  double width = 16;
  double height = 16;
  String type = "manhole";
  bool isActive = false;

  HitboxRect getHitbox() {
    return HitboxRect(x + 1, y + 2, width - 2, height - 3);
  }
}

class CoinEntity {
  double x = 0;
  double y = 0;
  bool isActive = false;
  bool isCollected = false;

  HitboxRect getHitbox() {
    return HitboxRect(x, y, 8, 8);
  }
}

class PowerUpEntity {
  double x = 0;
  double y = 0;
  PowerUpType type = PowerUpType.turbo;
  bool isActive = false;

  HitboxRect getHitbox() {
    return HitboxRect(x, y, 16, 16);
  }
}

class PassengerEntity {
  double x = 0;
  double y = 0;
  bool isLeftKerb = true;
  int typeIndex = 0;
  bool isPickedUp = false;
  int dropoffTargetMeter = 0;
  bool isActive = false;

  HitboxRect getHitbox() {
    return HitboxRect(x, y, 16, 24);
  }
}

class PedestrianEntity {
  double x = 0;
  double y = 0;
  bool isDodging = false;
  double dodgeOffset = 0;
  bool isActive = false;

  HitboxRect getHitbox() {
    final curX = isDodging ? x + dodgeOffset : x;
    return HitboxRect(curX + 3, y + 4, 10, 16);
  }
}

class PixelParticleEntity {
  double x = 0;
  double y = 0;
  double vx = 0;
  double vy = 0;
  int colorStepIndex = 0;
  double life = 0;
  double maxLife = 0.5;
  bool isActive = false;
}

class FloatingTextEntity {
  String text = "";
  double x = 0;
  double y = 0;
  double life = 0;
  double maxLife = 0.6;
  int colorArgb = 0;
  bool isActive = false;
}

File: app/src/main/java/com/example/components/GameEntities.kt
package com.example.components

import com.example.core.Constants
import com.example.core.PowerUpType

class RickshawPlayer {
    var lane: Int = 1
    var x: Float = Constants.LANE_CENTERS[1] - (Constants.RICKSHAW_WIDTH / 2f)
    var y: Float = Constants.RICKSHAW_Y
    var startX: Float = x
    var targetX: Float = x
    var isChangingLane: Boolean = false
    var laneChangeProgress: Float = 1f

    var lives: Int = 3
    var maxLives: Int = 3
    var invincibleTimerSec: Float = 0f

    var isTurbo: Boolean = false
    var turboTimeSec: Float = 0f

    var isShielded: Boolean = false

    var isMagnetActive: Boolean = false
    var magnetTimeSec: Float = 0f

    var nearMissCount: Int = 0

    fun startLaneChange(newLane: Int) {
        if (newLane in 0 until Constants.LANE_COUNT && newLane != lane) {
            lane = newLane
            startX = x
            targetX = Constants.LANE_CENTERS[lane] - (Constants.RICKSHAW_WIDTH / 2f)
            isChangingLane = true
            laneChangeProgress = 0f
        }
    }

    fun update(dt: Float) {
        if (isChangingLane) {
            laneChangeProgress += dt / Constants.LANE_CHANGE_DURATION
            if (laneChangeProgress >= 1f) {
                laneChangeProgress = 1f
                isChangingLane = false
                x = targetX
            } else {
                x = startX + (targetX - startX) * laneChangeProgress
            }
        }

        if (isTurbo) {
            turboTimeSec -= dt
            if (turboTimeSec <= 0f) {
                isTurbo = false
                turboTimeSec = 0f
            }
        }

        if (isMagnetActive) {
            magnetTimeSec -= dt
            if (magnetTimeSec <= 0f) {
                isMagnetActive = false
                magnetTimeSec = 0f
            }
        }

        if (invincibleTimerSec > 0f) {
            invincibleTimerSec -= dt
            if (invincibleTimerSec < 0f) invincibleTimerSec = 0f
        }
    }

    fun getHitbox(): HitboxRect {
        // 20x24 inset anchored to lower body
        return HitboxRect(x + 6f, y + 16f, 20f, 24f)
    }
}

data class HitboxRect(val x: Float, val y: Float, val w: Float, val h: Float) {
    fun intersects(other: HitboxRect): Boolean {
        return x < other.x + other.w &&
                x + w > other.x &&
                y < other.y + other.h &&
                y + h > other.y
    }
}

class ObstacleEntity {
    var x: Float = 0f
    var y: Float = 0f
    var width: Float = 16f
    var height: Float = 16f
    var type: String = "manhole"
    var isActive: Boolean = false

    fun getHitbox(): HitboxRect {
        return HitboxRect(x + 1f, y + 2f, width - 2f, height - 3f)
    }
}

class CoinEntity {
    var x: Float = 0f
    var y: Float = 0f
    var isActive: Boolean = false
    var isCollected: Boolean = false

    fun getHitbox(): HitboxRect {
        return HitboxRect(x, y, 8f, 8f)
    }
}

class PowerUpEntity {
    var x: Float = 0f
    var y: Float = 0f
    var type: PowerUpType = PowerUpType.TURBO
    var isActive: Boolean = false

    fun getHitbox(): HitboxRect {
        return HitboxRect(x, y, 16f, 16f)
    }
}

class PassengerEntity {
    var x: Float = 0f
    var y: Float = 0f
    var isLeftKerb: Boolean = true
    var typeIndex: Int = 0
    var isPickedUp: Boolean = false
    var dropoffTargetMeter: Int = 0
    var isActive: Boolean = false

    fun getHitbox(): HitboxRect {
        return HitboxRect(x, y, 16f, 24f)
    }
}

class PedestrianEntity {
    var x: Float = 0f
    var y: Float = 0f
    var isDodging: Boolean = false
    var dodgeOffset: Float = 0f
    var isActive: Boolean = false

    fun getHitbox(): HitboxRect {
        val curX = if (isDodging) x + dodgeOffset else x
        return HitboxRect(curX + 3f, y + 4f, 10f, 16f)
    }
}

class PixelParticleEntity {
    var x: Float = 0f
    var y: Float = 0f
    var vx: Float = 0f
    var vy: Float = 0f
    var colorStepIndex: Int = 0
    var life: Float = 0f
    var maxLife: Float = 0.5f
    var isActive: Boolean = false
}

class FloatingTextEntity {
    var text: String = ""
    var x: Float = 0f
    var y: Float = 0f
    var life: Float = 0f
    var maxLife: Float = 0.6f
    var colorArgb: Int = 0
    var isActive: Boolean = false
}
